nexus/src/app/instance.rs

```diff
@@ -2186,18 +2166,114 @@ pub(crate) async fn notify_instance_updated(
     }
 }
 
+/// Determines the disposition of a request to start an instance given its state
+/// (and its current VMM's state, if it has one) in the database.
+fn instance_start_allowed(
+    log: &slog::Logger,
+    instance: &DbInstance,
+    vmm: &Option<DbVmm>,
+) -> Result<InstanceStartDisposition, Error> {
+    // If the instance has an active VMM, there's nothing to start, but this
+    // disposition of this call (succeed for idempotency vs. fail with an
+    // error describing the conflict) depends on the state that VMM is in.
+    //
+    // If the instance doesn't have an active VMM, see if the instance state
+    // permits it to start.
+    if let Some(vmm) = vmm {
```

Comment:
I briefly wondered about whether the `if` where each branch is a `match` ought to be combined into one big `match (vmm.runtime.state, instance.runtime.nexus_state)`, but honestly, I'm not sure which is clearer...

Reply:
I also think this is a push (particularly if much of this routine is about to get replaced with a call to `effective_state` anyway) so have left it as-is.